ext/RMagick/rmimage.c

```diff
@@ -7159,7 +7159,7 @@ Image_gray_q(VALUE self)
     return has_attribute(self, (MagickBooleanType (*)(const Image *, ExceptionInfo *))SetImageGray);
```

Comment:
You can remove typecasting in here.

Reply:
I can't because `SetImageGray` requires an `Image *` instead.

Reply:
I wonder whether this PR is necessary...

Reply:
We don't need the typecasts at the other places. We only need it at this spot and that is why I decided to remove the others.

Reply:
I can't understand why need to remove typecast in this PR.

Reply:
I would like to remove it because it is useless. It would be the same as calling `foobar(Image*)` with `foobar((Image*)image)` when `image` is already a `Image*`.

Reply:
At least, looks for me that this PR need a function declared with the correct type for SetImageGray (), like

```diff
diff --git a/ext/RMagick/rmimage.c b/ext/RMagick/rmimage.c
index ef75ba32..62e670c8 100644
--- a/ext/RMagick/rmimage.c
+++ b/ext/RMagick/rmimage.c
@@ -7124,7 +7124,7 @@ Image_get_pixels(VALUE self, VALUE x_arg, VALUE y_arg, VALUE cols_arg, VALUE row
  * @return the result of attr_test.
  */
 static VALUE
-has_attribute(VALUE self, MagickBooleanType (attr_test)(const Image *, ExceptionInfo *))
+has_const_attribute(VALUE self, MagickBooleanType (attr_test)(const Image *, ExceptionInfo *))
 {
     Image *image;
     ExceptionInfo *exception;
@@ -7140,6 +7140,11 @@ has_attribute(VALUE self, MagickBooleanType (attr_test)(const Image *, Exception
     return r ? Qtrue : Qfalse;
 }
 
+static inline VALUE
+has_attribute(VALUE self, MagickBooleanType (attr_test)(Image *, ExceptionInfo *))
+{
+    return has_const_attribute(self, (MagickBooleanType (*)(const Image *, ExceptionInfo *))attr_test);
+}
 
 /**
  * Return true if all the pixels in the image have the same red, green, and blue
@@ -7156,10 +7161,10 @@ VALUE
 Image_gray_q(VALUE self)
 {
 #if defined(HAVE_SETIMAGEGRAY)
-    return has_attribute(self, (MagickBooleanType (*)(const Image *, ExceptionInfo *))SetImageGray);
+    return has_attribute(self, SetImageGray);
 #else
 #if defined(IMAGEMAGICK_GREATER_THAN_EQUAL_6_8_9)
-    return has_attribute(self, IsGrayImage);
+    return has_const_attribute(self, IsGrayImage);
 #else
     // For ImageMagick 6.7
     Image *image;
@@ -7174,7 +7179,7 @@ Image_gray_q(VALUE self)
         image->colorspace = RGBColorspace;
     }
 
-    ret = has_attribute(self, IsGrayImage);
+    ret = has_const_attribute(self, IsGrayImage);
     image->colorspace = colorspace;
     return ret;
 #endif
@@ -7195,7 +7200,7 @@ Image_gray_q(VALUE self)
 VALUE
 Image_histogram_q(VALUE self)
 {
-    return has_attribute(self, IsHistogramImage);
+    return has_const_attribute(self, IsHistogramImage);
 }
 
 
@@ -8938,7 +8943,7 @@ Image_monitor_eq(VALUE self, VALUE monitor)
 VALUE
 Image_monochrome_q(VALUE self)
 {
-    return has_attribute(self, IsMonochromeImage);
+    return has_const_attribute(self, IsMonochromeImage);
 }
 
 
@@ -9565,7 +9570,7 @@ Image_opaque_channel(int argc, VALUE *argv, VALUE self)
 VALUE
 Image_opaque_q(VALUE self)
 {
-    return has_attribute(self, IsOpaqueImage);
+    return has_const_attribute(self, IsOpaqueImage);
 }
 
 
@@ -9797,7 +9802,7 @@ Image_paint_transparent(int argc, VALUE *argv, VALUE self)
 VALUE
 Image_palette_q(VALUE self)
 {
-    return has_attribute(self, IsPaletteImage);
+    return has_const_attribute(self, IsPaletteImage);
 }
 
 
```

Reply:
(It need to rethink function names...)

Reply:
This PR only removes the unnecessary typecast. I think there should be a separate PR after this one if you want to add a method that IsImageGray can use without a typecast.

Reply:
OK, I see 😊